spec/validators/instrument_type_validator_spec.rb

```diff
@@ -0,0 +1,215 @@
+        end
+      end
+
+      it 'minimum or maximum number of plates' do
+        plates = build_list(:pacbio_plate, 2)
+
+        run = build(:pacbio_run, system_name: 'Sequel IIe', plates: [plates.first])
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates]).to be_empty
+
+        # Minimum
+        run = build(:pacbio_run, system_name: 'Sequel IIe', plates: [])
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates]).to include('must have at least 1 plate')
+
+        # Maximum
+        run = build(:pacbio_run, system_name: 'Sequel IIe', plates:)
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates]).to include('must have at most 1 plate')
+      end
+    end
+
+    context 'wells' do
+      it 'minimum or maximum number of wells' do
+        run = build(:pacbio_run, system_name: 'Sequel IIe', plates: [build(:pacbio_plate, well_count: 1)])
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages).to be_empty
+
+        # Minimum
+        run = build(:pacbio_run, system_name: 'Sequel IIe', plates: [build(:pacbio_plate, well_count: 0)])
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates].first).to include("plate #{run.plates.first.plate_number} wells must have at least 1 well")
+
+        # Maximum
+        run = build(:pacbio_run, system_name: 'Sequel IIe', plates: [build(:pacbio_plate, well_count: 100)])
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates].first).to include("plate #{run.plates.first.plate_number} wells must have at most 96 well")
+      end
+    end
+  end
+
+  describe 'when the instrument type is Revio' do
+    before do
+      create(:pacbio_smrt_link_version, name: 'v12_revio', default: true)
+    end
+
+    context 'plates' do
+      it 'required attributes' do
+        instrument_types['revio']['models']['plates']['validations']['required_attributes']['options']['required_attributes'].each do |attribute|
+          plate = build(:pacbio_plate)
+          run = build(:pacbio_run, system_name: 'Sequel IIe', plates: [plate])
+          plate.send("#{attribute}=", nil)
+          instrument_type_validator = described_class.new(instrument_types:)
+          instrument_type_validator.validate(run)
+          expect(run.errors.messages[:plates]).to include("plate #{plate.plate_number} #{attribute} can't be blank")
```

Comment:
Should this also have `sequencing_kit_box_barcode` can't be blank?

Reply:
It already does. It loops through the required attributes for each entity which is plate_number and sequencing_kit_box_barcode for plates as per the config. The message includes plate_number to distinguish which is a bit tricky because if the plate_number is not valid it will not print it but not a lot else I can do.